// @mos-connection/model/dist/mosTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IMOSScope = exports.stringifyMosType = exports.getMosTypes = void 0;
const MosString128 = require("./mosTypes/mosString128");
const MosDuration = require("./mosTypes/mosDuration");
const MosTime = require("./mosTypes/mosTime");
/**
 * Returns utility-functions for handling of MosTypes.
 * @example
 * const mosTypes = getMosTypes(true)
 * const myString128 = mosTypes.mosString128.create('Hello world')
 * const myString = mosTypes.mosString128.stringify(myString128)
 * @param strict If true, creating out-of-spec values will throw an error (Example: Creating longer-than-128-characters long MosString128).
 */
function getMosTypes(strict) {
    return {
        strict,
        mosString128: getMosType(MosString128, strict),
        mosDuration: getMosType(MosDuration, strict),
        mosTime: getMosType(MosTime, strict),
    };
}
exports.getMosTypes = getMosTypes;
/**
 * If value is a MosType, stringify it.
 * Throw error otherwise
 */
function stringifyMosType(value, mosTypes) {
    if (mosTypes.mosTime.is(value))
        return { isMosType: true, stringValue: mosTypes.mosTime.stringify(value) };
    if (mosTypes.mosDuration.is(value))
        return { isMosType: true, stringValue: mosTypes.mosDuration.stringify(value) };
    if (mosTypes.mosString128.is(value))
        return { isMosType: true, stringValue: mosTypes.mosString128.stringify(value) };
    return { isMosType: false };
}
exports.stringifyMosType = stringifyMosType;
function getMosType(mosType, strict) {
    return {
        create: (value) => mosType.create(value, strict),
        validate: (value) => mosType.validate(value, strict),
        stringify: (value) => mosType.stringify(value),
        valueOf: (value) => mosType.valueOf(value),
        is: (value) => mosType.is(value),
    };
}
var IMOSScope;
(function (IMOSScope) {
    IMOSScope["OBJECT"] = "OBJECT";
    IMOSScope["STORY"] = "STORY";
    IMOSScope["PLAYLIST"] = "PLAYLIST";
})(IMOSScope = exports.IMOSScope || (exports.IMOSScope = {}));
