// @mos-connection/helper/dist/mosModel/profile1/mosAck.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MOSAck = void 0;
const XMLBuilder = require("xmlbuilder");
const MosMessage_1 = require("../MosMessage");
const model_1 = require("@mos-connection/model");
const Utils_1 = require("../../utils/Utils");
class MOSAck extends MosMessage_1.MosMessage {
    /** */
    constructor(ack, strict) {
        super('lower', strict);
        this.ID = ack.ID;
        this.Status = ack.Status;
        this.Description = ack.Description;
        this.Revision = ack.Revision;
    }
    /** */
    get messageXMLBlocks() {
        const root = XMLBuilder.create('mosAck');
        (0, Utils_1.addTextElementInternal)(root, 'objID', this.ID, undefined, this.strict);
        (0, Utils_1.addTextElementInternal)(root, 'objRev', this.Revision, undefined, this.strict);
        (0, Utils_1.addTextElementInternal)(root, 'status', model_1.IMOSAckStatus[this.Status], undefined, this.strict);
        (0, Utils_1.addTextElementInternal)(root, 'statusDescription', this.Description, undefined, this.strict);
        return root;
    }
}
exports.MOSAck = MOSAck;
