// @mos-connection/model/dist/model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IMOSListMachInfoDefaultActiveXMode = exports.IMOSObjectPathType = exports.IMOSObjectAirStatus = exports.IMOSAckStatus = exports.IMOSObjectStatus = exports.IMOSObjectType = void 0;
var IMOSObjectType;
(function (IMOSObjectType) {
    IMOSObjectType["STILL"] = "STILL";
    IMOSObjectType["AUDIO"] = "AUDIO";
    IMOSObjectType["VIDEO"] = "VIDEO";
    IMOSObjectType["OTHER"] = "OTHER";
})(IMOSObjectType = exports.IMOSObjectType || (exports.IMOSObjectType = {}));
var IMOSObjectStatus;
(function (IMOSObjectStatus) {
    IMOSObjectStatus["NEW"] = "NEW";
    IMOSObjectStatus["UPDATED"] = "UPDATED";
    IMOSObjectStatus["MOVED"] = "MOVED";
    IMOSObjectStatus["BUSY"] = "BUSY";
    IMOSObjectStatus["DELETED"] = "DELETED";
    IMOSObjectStatus["NCS_CTRL"] = "NCS CTRL";
    IMOSObjectStatus["MANUAL_CTRL"] = "MANUAL CTRL";
    IMOSObjectStatus["READY"] = "READY";
    IMOSObjectStatus["NOT_READY"] = "NOT READY";
    IMOSObjectStatus["PLAY"] = "PLAY";
    IMOSObjectStatus["STOP"] = "STOP";
})(IMOSObjectStatus = exports.IMOSObjectStatus || (exports.IMOSObjectStatus = {}));
var IMOSAckStatus;
(function (IMOSAckStatus) {
    IMOSAckStatus["ACK"] = "ACK";
    IMOSAckStatus["NACK"] = "NACK";
})(IMOSAckStatus = exports.IMOSAckStatus || (exports.IMOSAckStatus = {}));
var IMOSObjectAirStatus;
(function (IMOSObjectAirStatus) {
    IMOSObjectAirStatus["READY"] = "READY";
    IMOSObjectAirStatus["NOT_READY"] = "NOT READY";
})(IMOSObjectAirStatus = exports.IMOSObjectAirStatus || (exports.IMOSObjectAirStatus = {}));
var IMOSObjectPathType;
(function (IMOSObjectPathType) {
    IMOSObjectPathType["PATH"] = "PATH";
    IMOSObjectPathType["PROXY_PATH"] = "PROXY PATH";
    IMOSObjectPathType["METADATA_PATH"] = "METADATA PATH";
})(IMOSObjectPathType = exports.IMOSObjectPathType || (exports.IMOSObjectPathType = {}));
var IMOSListMachInfoDefaultActiveXMode;
(function (IMOSListMachInfoDefaultActiveXMode) {
    IMOSListMachInfoDefaultActiveXMode["MODALDIALOG"] = "MODALDIALOG";
    IMOSListMachInfoDefaultActiveXMode["MODELESS"] = "MODELESS";
    IMOSListMachInfoDefaultActiveXMode["CONTAINED"] = "CONTAINED";
    IMOSListMachInfoDefaultActiveXMode["TOOLBAR"] = "TOOLBAR";
})(IMOSListMachInfoDefaultActiveXMode = exports.IMOSListMachInfoDefaultActiveXMode || (exports.IMOSListMachInfoDefaultActiveXMode = {}));
